src/ip.rs

```diff
@@ -1,43 +1,44 @@
 use crate::Error;
-use ark_ec::{msm::VariableBaseMSM, AffineCurve, PairingEngine};
-use ark_ff::PrimeField;
-use ark_std::{cfg_iter, vec::Vec};
-use rayon::prelude::*;
+use ark_ec::{
+    pairing::{MillerLoopOutput, Pairing, PairingOutput},
+    AffineRepr, VariableBaseMSM,
+};
+use ark_std::vec::Vec;
 
-pub(crate) fn pairing_miller_affine<E: PairingEngine>(
+pub(crate) fn pairing_miller_affine<E: Pairing>(
     left: &[E::G1Affine],
     right: &[E::G2Affine],
-) -> Result<E::Fqk, Error> {
+) -> Option<MillerLoopOutput<E>> {
```

Comment:
Can we keep that a result to make it similar to the other functions there ?

Reply:
Yes, sure! Makes sense, done!